php/libraries/ICandidate.class.inc

```diff
@@ -0,0 +1,319 @@
+     */
+    public function getEntityType(): string;
+
+    /**
+     * Returns the list of visit labels w/ corresponding SessionID as key
+     *
+     * @return array<int,string> Associative array in the format
+     *                              SessionID=>Visit_label
+     */
+    public function getListOfVisitLabels(): array;
+
+    /**
+     * Generates list of valid subprojects for a the selected Candidate's
+     * project
+     *
+     * @return array<int,int> Array of the form SubprojectID => SubprojectID
+     *
+     * @throws DatabaseException
+     */
+    public function getValidSubprojects(): array;
+
+    /**
+     * Returns the Visit_label for a subject's first TimePoint.
+     *
+     * @return string The visit label for the candidate's first visit.
+     *                  Returns an empty string if one doesn't exist.
+     */
+    public function getFirstVisit(): string;
+
+    /**
+     * Returns the next visit number that should be created for this
+     * candidate.
+     *
+     * @return int The next visit number
+     */
+    public function getNextVisitNo() : int;
+
+    /**
+     * Returns list of consents and their respective statuses for this candidate
+     *
+     * @return array List of consents and their associated values for this candidate
+     *               The keys of the arrays are the IDs of the consents
+     */
+    public function getConsents(): array;
+
+    /**
+     * Gets the participant_status options from participant_status_options
+     * table. Used to populate LorisForm.
+     *
+     * @return string[]
+     */
+    static function getParticipantStatusOptions(): array;
+
+    /**
+     * Gets the participant_status options suboptions from participant_status_options
+     *
+     * @param int $parentID parent ID of the participant status option
+     *
+     * @return string[] Options array suitable for use in LorisForm select
+     *               element
+     */
+    public static function getParticipantStatusSubOptions(int $parentID): array;
+
+    /**
+     * Sets generic data in this class to $newData.
+     * Manages the parameter_candidate table so that projects have a way to add
+     * whatever data they want which is outside the scope of LORIS to their
+     * candidates.
+     *
+     * @param array $newData Key-value pair(s) of the field(s) to change and the
+     *                              new value(s).
+     *
+     * @return void
+     */
+    public function setData(array $newData): void;
+
+    /**
+     * Validate's that a PSCID matches the format defined by project
+     * config
+     *
+     * @param string $PSCID The PSCID to validate
+     *
+     * @return bool
+     */
+    public static function validatePSCID(string $PSCID): bool;
+
+    /**
+     * This is a function to supply data about the Candidate object to the
+     * front-end. Many Smarty templates refer to the specific column names
+     * and aliases used below.
```

Comment:
I can see why it would be useful but I can't see where it is used right now. Do you a an example @johnsaigle ?

Reply:
This is what `getData()` currently does. An example:

https://github.com/aces/Loris/blob/01c743af6721b0912abebebd1f742b30ec5c2ef0/src/Middleware/UserPageDecorationMiddleware.php#L77

There are many others though. Essentially an array representation of Candidate (the `$_candidateInfo` class variable) is passed directly to the template for rendering.

This isn't ideal but I think an interface will need to support this for legacy reasons until Smarty is fully removed. At that point it can be deprecated and eventually removed.